File: frontend/electricity-market-prediction/src/types/stock.ts
/**
 * @fileoverview 當日漲停 + 分點券商型別。
 *
 * 分類欄位兩種來源：
 *   - concept        ── 子產業細分（MoneyDJ）
 *   - concept_reason ── 基礎分類（TWSE 官方產業別）
 */

export type Market = 'twse' | 'tpex';
export type ClassifyMode = 'base' | 'sub';

export interface BrokerEntry {
    broker: string;
    buy: number;
    sell: number;
    net: number;
}

export interface DailyStock {
    code: string;
    name: string;
    market: Market;
    close: number;
    changePercent: number;
    volume: number;

    concept: string;
    concept_reason: string;

    foreign: number;
    trust: number;
    dealer: number;

    brokers: BrokerEntry[];  // 保留向下相容；新流程改用 BrokerSnapshotResponse
}

export interface SectorOption { name: string; count: number; }
export interface MarketBreakdown { twse: number; tpex: number; }

export interface DailyLimitUpResponse {
    date: string;
    updatedAt: string;
    total: number;
    breakdown: MarketBreakdown;
    baseSectors: SectorOption[];
    subSectors: SectorOption[];
    stocks: DailyStock[];
}

// ── 分點券商 ──────────────────────────────────────────────
export type BrokerStatus = 'pending' | 'ok' | 'failed';

export interface BrokerEntryOut {
    broker_code: string;
    broker_name: string;
    net: number;
    buy: number;
    sell: number;
    buy_avg?: number | null;
    sell_avg?: number | null;
    rank_in_buy?: number | null;
    rank_in_sell?: number | null;
}

// ── 可用歷史日期 ──────────────────────────────────────────
export interface AvailableDateRow { date: string; total: number; }
export interface AvailableDatesResponse {
    /** 已 snapshot 的日期清單 (desc by date) */
    dates: AvailableDateRow[];
    /** 後端認定的今日(Asia/Taipei) ISO 字串 — 前端用以區分 live vs 歷史 */
    today: string;
}

export interface BrokerSnapshotResponse {
    code: string;
    name: string;
    market: string;
    trade_date?: string | null;
    status: BrokerStatus;
    fetched_at?: string | null;
    error?: string | null;
    open?: number | null;
    high?: number | null;
    low?: number | null;
    close?: number | null;
    total_records: number;
    total_brokers: number;
    buyTop: BrokerEntryOut[];
    sellTop: BrokerEntryOut[];
    all: BrokerEntryOut[];
}

File: backend/app/schemas/stock.py
"""當日漲停（daily）API schemas。

分類欄位兩來源（沿用前端 / 需求命名）：
  - concept        ── 子產業（MoneyDJ）
  - concept_reason ── 基礎分類（TWSE 官方產業別）
"""
from __future__ import annotations

from pydantic import BaseModel, Field


class BrokerEntry(BaseModel):
    broker: str
    buy: int = 0
    sell: int = 0
    net: int = 0


class DailyStock(BaseModel):
    code: str
    name: str
    market: str = Field(description="twse / tpex")
    close: float
    changePercent: float
    volume: int
    concept: str = ""            # 子產業（MoneyDJ）
    concept_reason: str = ""     # 基礎分類（TWSE）
    foreign: int = 0
    trust: int = 0
    dealer: int = 0
    brokers: list[BrokerEntry] = Field(default_factory=list)


class SectorOption(BaseModel):
    name: str
    count: int


class MarketBreakdown(BaseModel):
    twse: int = 0
    tpex: int = 0


class DailyLimitUpResponse(BaseModel):
    date: str
    updatedAt: str
    total: int
    breakdown: MarketBreakdown = Field(default_factory=MarketBreakdown)
    baseSectors: list[SectorOption] = Field(default_factory=list)
    subSectors: list[SectorOption] = Field(default_factory=list)
    stocks: list[DailyStock] = Field(default_factory=list)
